src/iterable/mod.rs

```diff
@@ -0,0 +1,106 @@
+//! A base library for interfacing with streams of vectors and matrices.
+//!
+//! This library presents the abstraction layer for the _streaming model_.
+//! Essentially, it provides a set of handy utilities as a wrapper around
+//! iterators.
+
+/// The trait [`Iterable`] represents a streamable object that can produce
+/// an arbitrary number of streams of length [`Iterable::len`](Iterable::len).
+///
+/// An Iterable is pretty much like an [`IntoIterator`] that can be copied over
+/// and over, and has an hint of the length.
+/// In other words, these two functions are pretty much equivalent.
```

Comment:
This sounds like the entire Iterable trait is not necessary, and it is the duty for Gemini to not use Iterable but instead to use the standard library one.

What is the main difference that Iterable provides but not by `IntoIterator`?

Reply:
Fair point: the main difference is that we need to iter over the same stream multiple times, by design.
This means that already we require `IntoIterator + Copy` or `IntoIterator + Clone` if we want to go down that avenue. 

From there, it's pretty much the same, except that now to get the length of a stream it becomes a bit more verbose, and the length function might disagree with the actual length (this is needed for instance when streaming a matrix).
Here's a bit of an overview of how the code might become bigger: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c7d9d2fca7d1d308601bd5b32d770409

To be fair, on the one hand, the downside is that `Iterable::iter` will have to copy all relevant data from the initial structure (until https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md hits the streets). On the other hand, for the long-term vision we have a trait that can be extended with other functions on the top of normal iterables, and that we skip having to type things like `vec.into_iter().len()`.